### DISCUSSION TITLE
Access VERCEL_URL in Client

### DISCUSSION THREAD

**Discussion by danlutz** [vercel/next.js#15424] [Help] [answered]
Hi there,

We are hosting our Next.js app on Vercel, which is awesome. 
I was wondering how we can access `process.env.VERCEL_URL`, which is populated by Vercel at build time, in our client side code. It works perfectly in api routes, but on the client side it's `undefined`. Is there a way to work around this? I guess it's not working because it's not prefixed with `NEXT_PUBLIC_`

For reference:
https://vercel.com/docs/v2/build-step#system-environment-variables

Thanks alot!

**Comment by Timer** [comment, marked as answer]
Edit: [Please see this comment](https://github.com/vercel/next.js/discussions/15424#discussioncomment-269353).

---

You can create a `.env.production` file with the following inside of it:
```
# .env.production
NEXT_PUBLIC_VERCEL_URL=${VERCEL_URL}
```

This will then let you reference `process.env.NEXT_PUBLIC_VERCEL_URL` in your app.

**Reply by danlutz** [reply]
So i should commit this env file to my repository then?

**Reply by Timer** [reply]
Yes!

**Reply by danlutz** [reply]
Thank you! :)

**Reply by tettoffensive** [reply]
Can you just do this on the dashboard by setting the value as `${VERCEL_URL}`?

**Reply by lukeshumard** [reply]
For any new users coming here, this solution requires you to have `VERCEL_URL` set in your environment variables on Vercel. This is what populates the `VERCEL_URL` variable at build time, as it is not included automatically. This was a gotcha for me when setting it up, so I wanted to share.

**Reply by bscaspar** [reply]
@lukeshumard to confirm, if I want to use the `VERCEL_URL` env var, I need to go into my dashboard and create a new 'Reference to System Environment Variable' type var with the name = `VERCEL_URL` (or anything I want) and value is the reference to `VERCEL_URL`?

**Reply by lukeshumard** [reply]
Edit for Dec 2021: Please see [this comment](https://github.com/vercel/next.js/discussions/15424#discussioncomment-1752536)
___
@bscaspar This has since changed on Vercel from when I wrote my reply above. Vercel's environment variables UI now allows you to set a reference to a System Environment Variable with any name.

This will work with the suggested answer above.
<img width="781" alt="Screen Shot 2021-01-08 at 2 39 12 PM" src="https://user-images.githubusercontent.com/472555/104028899-2ff42700-51c1-11eb-949e-6fa007a700a8.png">

And this works without needing to also add a variable called `VERCEL_URL` in your project.
<img width="791" alt="Screen Shot 2021-01-08 at 2 40 16 PM" src="https://user-images.githubusercontent.com/472555/104028999-50bc7c80-51c1-11eb-8edc-1178f8362efa.png">

**Reply by activenode** [reply]
This is not true anymore. At least it doesn't show in the UI. Instead this: https://github.com/vercel/next.js/discussions/15424#discussioncomment-1752536

**Comment by bitcoder** [comment]
I have tried all possible combinations and I was unable to get this working in client side.
I tried:

- enabling "Automatically expose system environment variables" and in the code reference process.env.VERCEL_URL
- create a .env.production file with the content "NEXT_PUBLIC_VERCEL_URL=${VERCEL_URL}".. and  in the code  reference process.env.VERCEL_URL
- create an environment variabled named VERCEL_URL, mapped to the VERCEL_URL system env, and in the code reference process.env.VERCEL_URL
- create an environment variabled named NEXT_PUBLIC VERCEL_URL,mapped to the VERCEL_URL system env,  and in the code reference process.env.NEXT_PUBLIC_VERCEL_URL


Not sure what I can do more. Can anyone help please?

**Reply by bscaspar** [reply]
Have you tried to set this in your project settings on the Vercel site? If you expand the above answers you will see Luke's response to my question with a screenshot demoing what to do that should work.

**Reply by bitcoder** [reply]
yes @bscaspar, if you mean the answer on https://github.com/vercel/next.js/discussions/15424#discussioncomment-269353 ,  but unfortunately it didn't work.

**Reply by bitcoder** [reply]
I got it worked. 
In project settings I created an environment variabled named NEXT_PUBLIC VERCEL_URL,mapped to the VERCEL_URL system env, and in the code obtained it using process.env.NEXT_PUBLIC_VERCEL_URL.
Not sure what happened..
Thanks!

**Reply by bitcoder** [reply]
but it seems I'm blocked on the VERCEL_ENV.. isn't it available on client side?
My challenge is to get the VERCEL_URL on client side but having it pointing to the domain when it's in production... currently it always gives me the temporary environment. thus, I was thinking on implementing some logic that would look at VERCEL_ENV and check if it's production deployment or not

**Reply by bscaspar** [reply]
None of the system env vars are available client side, you need to do the same thing you did for `VERCEL_URL` by creating a `NEXT_PUBLIC_VERCEL_ENV` variable that points to `VERCEL_ENV` in order to access it client-side.

**Reply by bitcoder** [reply]
Right, i figured that out yesterday :)  but as it seems I dont need to create a "NEXT_PUBLIC_VERCEL_ENV" explicitly in the UI as VERCEL_ENV is available by default (if I have the "expose vars" setting checked) and is accessible by referring to it as NEXT_PUBLIC_VERCEL_ENV.

**Reply by bscaspar** [reply]
Good to know!

**Reply by chris-stytch** [reply]
> Right, i figured that out yesterday :) but as it seems I dont need to create a "NEXT_PUBLIC_VERCEL_ENV" explicitly in the UI as VERCEL_ENV is available by default (if I have the "expose vars" setting checked) and is accessible by referring to it as NEXT_PUBLIC_VERCEL_ENV.

@bitcoder do you mean that so long as you have "Automatically expose System Environment Variables" checked, you can directly call `process.env.VERCEL_URL` client-side? I can't seem to get any combination of configuration mentioned in this thread to work.

**Reply by prichodko** [reply]
Vercel automatically makes predefined system environment variables available during build step with the correct prefix. Just scroll a bit down on this [page](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables). You can simply use `NEXT_PUBLIC_VERCEL_URL` to use it on the client side.

**Comment by bitcoder** [comment]
Well, this was some time ago and I don't remember exactly all the details.
Anyway, this is the logic that I use nowadays in a lib/contants.js which I then import from other places

```
const vercel_public_domain = process.env.NEXT_PUBLIC_DOMAIN

const node_prod_env = process.env.NODE_ENV === 'production'

// if Vercel prod deployment, then use main domain; else use temporary domain unless you're on dev
export const DOMAIN =  vercel_public_domain ? vercel_public_domain : (node_prod_env ? `https://${process.env.NEXT_PUBLIC_VERCEL_URL}` : 'http://127.0.0.1:3000')
```

**Comment by prichodko** [comment]
The correct answer today is to directly use `NEXT_PUBLIC_VERCEL_URL`, which is automatically available during build step. You can read more [here](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables) – scroll to the section addressing frameworks.

**Reply by smakosh** [reply]
Don't forget to append `https://${NEXT_PUBLIC_VERCEL_URL}` where that env variable is used

**Reply by genaroibc** [reply]
That's VERY important...

**Reply by m3c-ode** [reply]
If the `process.env.NEXT_PUBLIC_VERCEL_URL` const is called in our program, do we need to add it in the Env Variables of the project in Vercel's UI? Or is it automatic? So far, my app still makes calls to `https://undefined/api...`

**Reply by smakosh** [reply]
It's a default env variable, make sure your program is loading the env variables properly

**Reply by m3c-ode** [reply]
When I look at the deployment console's log, I can properly log the `process.env.NEXT_PUBLIC_VERCEL_URL`, but if I click on a button that's supposed to make a request to this URL, it still appears undefined... not sure why

**Reply by prichodko** [reply]
@m3c-ode can you show the code of your button?

**Reply by jacob-mako** [reply]
I saw this when I did the following:

```typescript
export const getBaseUrl = () => {
  if (process.env.NODE_ENV === 'development') {
    return `${process.env.NEXT_PUBLIC_SITE_URL}`
  }

  return `https://${process.env.NEXT_PUBLIC_VERCEL_URL}`
}
const baseUrl = getBaseUrl()
const pathToAssetUrl = `${baseUrl}/${publicDirFileName}`
console.log(`Resolved deployment URL to: ${pathToAssetUrl}`)
```

**Comment by tresorama** [comment]
Vercel docs has two pages that we should read to understand what Vercel does automatically for Env Vars

https://vercel.com/docs/projects/environment-variables/framework-environment-variables
https://vercel.com/docs/projects/environment-variables/system-environment-variables

Initially I found only the "system" one on Google.

**Comment by tresorama** [comment]
When we deploy a Next.js app to Vercel, the app is reachable at two URLs.
```bash
# 1 - the production url for the project that doesn't change with subsequent git commit and deploys
https://my-site.vercel.app 

# 2- the url of the deploy of this git commit
https://my-site-bmprob6pr.vercel.app/ 
```

Because of that if we log Vercel system variables **inside the frontend** console we get

```js
console.log({
  NEXT_PUBLIC_VERCEL_ENV: process.env.NEXT_PUBLIC_VERCEL_ENV,
  NEXT_PUBLIC_VERCEL_URL: process.env.NEXT_PUBLIC_VERCEL_URL,
  NEXT_PUBLIC_VERCEL_BRANCH_URL: process.env.NEXT_PUBLIC_VERCEL_BRANCH_URL,
  NEXT_PUBLIC_VERCEL_PROJECT_PRODUCTION_URL: process.env.NEXT_PUBLIC_VERCEL_PROJECT_PRODUCTION_URL,
});

{
  "NEXT_PUBLIC_VERCEL_ENV": "production",
  "NEXT_PUBLIC_VERCEL_URL": "my-site-bmprob6pr.vercel.app",
  "NEXT_PUBLIC_VERCEL_BRANCH_URL": "my-site-git-d31e9f-user-projects.vercel.app",
  "NEXT_PUBLIC_VERCEL_PROJECT_PRODUCTION_URL": "my-site.vercel.app"
}
```

This is how I get the app url in my code
```ts
const getBaseUrl = () => {
  if (process.env.NODE_ENV ===  "development") {
    return "http://localhost:3000";
  }

  const isVercel = process.env.NEXT_PUBLIC_VERCEL_PROJECT_PRODUCTION_URL !== undefined;
  if (isVercel) return `https://${process.env.NEXT_PUBLIC_VERCEL_PROJECT_PRODUCTION_URL}`;

  return "url not resolved";
};
```
...but with that I have the same output in both domains (1 and 2).


## There is a way to determine from **frontend** if we are in production URL (1) or commit deploy URL (2) ?

**Reply by grrowl** [reply]
`document.location.hostname`
